### DISCUSSION TITLE
Should MolToSmiles/Inchi respect enhanced stereo of singleton &1s?

### DISCUSSION THREAD

**Discussion by pechersky** [rdkit/rdkit#6908] [FAQ]
I'm asking about 1-atom `STERAC` groups, like the following:
<img width="313" alt="image" src="https://github.com/rdkit/rdkit/assets/5342086/983a337d-f23e-4e8f-9c66-9c797a6ded5e">

```

  ChemDraw11132320552D

  0  0  0     0  0              0 V3000
M  V30 BEGIN CTAB
M  V30 COUNTS 4 3 0 0 0
M  V30 BEGIN ATOM
M  V30 1 C -0.714471 -0.618750 0.000000 0
M  V30 2 C -0.000000 -0.206250 0.000000 0
M  V30 3 Cl 0.714471 -0.618750 0.000000 0
M  V30 4 F -0.000000 0.618750 0.000000 0
M  V30 END ATOM
M  V30 BEGIN BOND
M  V30 1 1 2 1
M  V30 2 1 2 3
M  V30 3 1 2 4
M  V30 END BOND
M  V30 END CTAB
M  END
$$$$

  ChemDraw11132320552D

  0  0  0     0  0              0 V3000
M  V30 BEGIN CTAB
M  V30 COUNTS 4 3 0 0 0
M  V30 BEGIN ATOM
M  V30 1 C -0.714471 -0.618750 0.000000 0
M  V30 2 C -0.000000 -0.206250 0.000000 0
M  V30 3 Cl 0.714471 -0.618750 0.000000 0
M  V30 4 F -0.000000 0.618750 0.000000 0
M  V30 END ATOM
M  V30 BEGIN BOND
M  V30 1 1 2 1 CFG=1
M  V30 2 1 2 3
M  V30 3 1 2 4
M  V30 END BOND
M  V30 BEGIN COLLECTION
M  V30 MDLV30/STERAC1 ATOMS=(1 2)
M  V30 END COLLECTION
M  V30 END CTAB
M  END
$$$$

  ChemDraw11132320552D

  0  0  0     0  0              0 V3000
M  V30 BEGIN CTAB
M  V30 COUNTS 4 3 0 0 0
M  V30 BEGIN ATOM
M  V30 1 C -0.714471 -0.618750 0.000000 0
M  V30 2 C -0.000000 -0.206250 0.000000 0
M  V30 3 Cl 0.714471 -0.618750 0.000000 0
M  V30 4 F -0.000000 0.618750 0.000000 0
M  V30 END ATOM
M  V30 BEGIN BOND
M  V30 1 1 2 1
M  V30 2 1 2 3
M  V30 3 1 2 4 CFG=1
M  V30 END BOND
M  V30 BEGIN COLLECTION
M  V30 MDLV30/STERAC1 ATOMS=(1 2)
M  V30 END COLLECTION
M  V30 END CTAB
M  END
$$$$
```

`Chem.MolToSmiles` for the latter two shows a stereocenter. `Chem.MolToInchi` also shows stereoblocks. Is this expected/desired?

**Comment by bp-kelley** [comment]
Smiles has no knowledge of extended stereo groups, CXSMiles, however, does:


```
data = """"<above>"""
for m in Chem.ForwardSDMolSupplier(BytesIO(data.encode())):
    print(Chem.MolToCXSmiles(m))
```

results in:

```
CC(F)Cl |(-0.714471,-0.61875,;0,-0.20625,;0,0.61875,;0.714471,-0.61875,)|
C[C@H](F)Cl |(-0.714471,-0.61875,;0,-0.20625,;0,0.61875,;0.714471,-0.61875,),wU:1.0,&1:1|
C[C@H](F)Cl |(-0.714471,-0.61875,;0,-0.20625,;0,0.61875,;0.714471,-0.61875,),wU:1.1,&1:1|
```

**Comment by bp-kelley** [comment]
This is the ChemAxon extended smiles (btw):

https://docs.chemaxon.com/display/docs/chemaxon-extended-smiles-and-smarts-cxsmiles-and-cxsmarts.md

**Comment by pechersky** [comment]
Right. I understand that SMILES doesn't understand extended stereo groups. My question is regarding how much "canonicalization" I should be expecting that RDKit does before/during `MolToSmiles` when putting in a molecule with such `&1`s. It already does some type of canonicalization over atom order (and thus CW/CCW inference). And during `MolToInchi`, there is an internal tautomer canonicalization path.

**Comment by bp-kelley** [comment]
It really does sound like CXSmiles is what you want here, it does canonicalize the "&"s.


```
from rdkit.Chem.RegistrationHash import GetStereoTautomerHash

for i,m in enumerate(Chem.ForwardSDMolSupplier(BytesIO(data.encode()))):
    m.RemoveAllConformers() # remove coords
    print("smi", Chem.MolToSmiles(m))

for i,m in enumerate(Chem.ForwardSDMolSupplier(BytesIO(data.encode()))):
    m.RemoveAllConformers() # remove coords
    print("cxsmi", Chem.MolToCXSmiles(m))

```

Results in 

```
0 smi  : CC(F)Cl
1 smi  : C[C@@H](F)Cl
2 smi  : C[C@@H](F)Cl

0 cxsmi: CC(F)Cl
1 cxsmi: C[C@H](F)Cl |&1:1|
2 cxsmi: C[C@H](F)Cl |&1:1|
```

**Comment by bp-kelley** [comment]
Note, you might want to remove coordinates when outputting cxsmiles for canonicalization purposes.

**Reply by kienerj** [reply]
Wouldn't the cleaner way be to use the appropriate flag, CX_ALL_BUT_COORDS? 

    Chem.MolToCXSmiles(m, flags=CXSmilesFields.CX_ALL_BUT_COORDS)

**Reply by bp-kelley** [reply]
Sometimes I can’t keep the whole API in my head :)

And yes, this is more clear.

**Comment by pechersky** [comment]
Do you think there is a way to get a "spirit of the law" way to say these 3 molecules represent the same mixture? One option could be comparing the sets of SMILES strings that the enumerator produces.

**Comment by bp-kelley** [comment]
I think that currently, generating the isomers is probably the best way, we do have. Registration Hash but it currently has different results for 0 and (1,2)

```
# enumerate isomers smiles into set()
... print(".".join(sorted(isomer_smi))

0 isomers C[C@@H](F)Cl.C[C@H](F)Cl
1 isomers C[C@@H](F)Cl.C[C@H](F)Cl
2 isomers C[C@@H](F)Cl.C[C@H](F)Cl
```

Currently we can't add 'missing' & stereo groups.   That might be a nice feature, actually.

**Comment by pechersky** [comment]
This works for singleton `&1`, but doesn't work for singleton `or1`, which are supposed to indicate discernable molecules:
```
print(Chem.MolToCXSmiles(Chem.MolFromSmiles("C[C@H](F)Cl |o1:1|")))  # C[C@H](F)Cl |o1:1|
print(Chem.MolToCXSmiles(Chem.MolFromSmiles("C[C@@H](F)Cl |o1:1|")))  # C[C@H](F)Cl |o1:1|

for smiles in ["C[C@H](F)Cl |o1:1|", "C[C@@H](F)Cl |o1:1|"]:
  mol = Chem.MolFromSmiles(smiles)
  isomers = tuple(EnumerateStereoisomers(mol))
  print(".".join(sorted(Chem.MolToSmiles(x, isomericSmiles=True) for x in isomers)))

# C[C@@H](F)Cl.C[C@H](F)Cl
# C[C@@H](F)Cl.C[C@H](F)Cl
```

**Reply by bp-kelley** [reply]
I think the missing piece is the number of isomers “in the sample”.

&1 == two isomers 
Or1 == one isomer out of 2

I don’t think we have a direct way of calculating this.

What, exactly, are you trying to do?

**Comment by bp-kelley** [comment]
However, to answer your question, SMILES was never intended to solve this particular problem that extended stereo chemistry solves, so it should not consider it in canonicalization.   While inchi does normalize some tautomers, it also has the same issue regarding extended stereochemistry.

**Comment by bp-kelley** [comment]
Loosely I consider the smiles notation to indicate what a molecule IS and extended stereo chemistry to indicate what a molecule could be.

**Comment by pechersky** [comment]
I'm trying to understand which string representation of a molecule throws or keeps information in what way. @coleb suggested I contribute the following two functions:
`Chem.EnhancedStereoToTraditionalStereo`, and back again `Chem.TraditionalStereoToEnhancedStereo` that did the proper conversion of singleton &1 and abs tags

**Reply by bp-kelley** [reply]
Turns out, I’m currently writing one of them.


```
smi='C[C@H](F)Cl' assignStereoGroups => 'C[C@H](F)Cl |a:1|'
smi='C[CH](F)Cl' assignStereoGroups => 'C[C@H](F)Cl |&1:1|'
smi='C[C@@H](F)Cl |&1:1|' assignStereoGroups => 'C[C@H](F)Cl |&1:1|'
```

**Reply by bp-kelley** [reply]
I’m not sure this can be done both ways.

**Reply by coleb** [reply]
@bp-kelley, have an example that can’t round trip through the two conventions? We haven’t found one.

**Reply by bp-kelley** [reply]
What direction are we going in?  Everything in smiles can be represented in CXSmiles, but

```
C[C@H](F)Cl |or1:1|
```

Has no direct smiles representation.


Additionally

```
C[C@H](F)Cl.C[C@H](F)Cl |$1:1,&1:5|
```

has no smiles representation.

**Reply by bp-kelley** [reply]
Also, it is important to think about "sample isomers" versus "potential isomers"

```
C[C@H](F)Cl |&1:1|
```

is one sample with two isomers

```
C[C@H](F)Cl |or1:1|
```

can be two samples with one isomer each.

**Reply by bp-kelley** [reply]
I.e. if you were splitting on a chiral column, how many peaks should be in the sample.

**Reply by bp-kelley** [reply]
Here are some stupid examples:

```
smi='C[C@H](F)Cl' assignStereoGroups => 'C[C@H](F)Cl |a:1|'  (1, 1), possible_isomers=1 isomers_in_sample=1
smi='C[CH](F)Cl' assignStereoGroups => 'C[C@H](F)Cl |&1:1|'  (2, 2), possible_isomers=2 isomers_in_sample=2
smi='C[C@@H](F)Cl |&1:1|' assignStereoGroups => 'C[C@H](F)Cl |&1:1|'  (2, 2), possible_isomers=2 isomers_in_sample=2
smi='C[C@@H](F)Cl |o1:1|' assignStereoGroups => 'C[C@H](F)Cl |o1:1|'  (2, 1), possible_isomers=2 isomers_in_sample=1
smi='C[C@@H](F)Cl.C[C@@H](F)Cl |&1:1,&2:5|' assignStereoGroups => 'C[C@H](F)Cl.C[C@H](F)Cl |&1:1,&2:5|'  (4, 4), possible_isomers=4 isomers_in_sample=4
smi='C[C@@H](F)Cl.C[C@@H](F)Cl |&1:1,o2:5|' assignStereoGroups => 'C[C@H](F)Cl.C[C@H](F)Cl |o1:5,&1:1|'  (4, 2), possible_isomers=4 isomers_in_sample=2
```

**Reply by bp-kelley** [reply]
Which boils down to "what problem are you trying to solve."

I'm close to making an EnumerateSampleIsomers which outputs all the isomers you have in each samples or lot.  EnumerateStereoisomers currently only outputs the isomers you could be regardless of sample/lot.

It really sounds like you all are trying to make a registration system to mix assay results from chemical lots (could be many isomers) with things like SPR or computational results which are tied to a particular isomer.  

I'll just say, we should talk, since I'm really close to being able to do this :)

**Reply by pechersky** [reply]
Let's find a zoom call time =) feel free to email me to set it up

**Reply by coleb** [reply]
Our thinking was the conversion functions `Chem.EnhancedStereoToTraditionalStereo` and `Chem.TraditionalStereoToEnhancedStereo` would focus on only converting the two points of ambiguity between the two conventions. Specifically these two: 

```
1. 'C[C@H](F)Cl' <=> 'C[C@H](F)Cl |a:1|'
2. 'C[CH](F)Cl' <=> 'C[C@H](F)Cl |&1:1|'
```

1. add/remove 'abs' tag
2. add/remove '&1' when only on a single atom, flattening the bond directions if necessary, removing CW/CCW if necessary

We would leave multi-atom 'and' groups alone. We would leave or groups alone. 

We should definitely chat! Think we may be attacking at different angles.

**Reply by bp-kelley** [reply]
Yeah, my goal is given a structure "as drawn" how many samples does it describe  and in those samples, how many isomers are in them.  So I'm really only going in one direction.

I think round tripping only works for really simple structures which is why I'm describing them as isomer sets to remove ambiguity.

Also, once you have the extended stereo chem set, you can use Chem.RegistrationHash to make your canonical key.

**Reply by bp-kelley** [reply]
I'll see if I can get my registration/normalization stuff up on github we can use that as a straw man,.

**Reply by kienerj** [reply]
> Our thinking was the conversion functions Chem.EnhancedStereoToTraditionalStereo and Chem.TraditionalStereoToEnhancedStereo would focus on only converting the two points of ambiguity between the two conventions. Specifically these two:

>1. 'C[C@H](F)Cl' <=> 'C[C@H](F)Cl |a:1|'
>2. 'C[CH](F)Cl' <=> 'C[C@H](F)Cl |&1:1|'

>    add/remove 'abs' tag
>    add/remove '&1' when only on a single atom, flattening the bond directions if necessary, removing CW/CCW if necessary

>We would leave multi-atom 'and' groups alone. We would leave or groups alone.

For me this makes a ton of sense and would be useful for the exact reasons described.

**Reply by bp-kelley** [reply]
Here is a straw man for you all:


```
from rdkit import Chem
import logging

def determineStereoGroups(mol, determine_unassigned_stereo=True):
  """This converts unassigned stereo to stereo groups

   Args:
   :mol: Molecule to convert
   :determine_unassigned_stereo: if True generate stereo groups from unassigned stereo [True]

  Returns:
   New molecule with assigned stereo groups
  """
  rwmol = Chem.RWMol(mol)
  Chem.FindPotentialStereoBonds(rwmol)
  
  min_idx = 0
  atoms_in_stereo_groups = set()
  for sg in rwmol.GetStereoGroups():
    for atom in sg.GetAtoms():
      atoms_in_stereo_groups.add(atom.GetIdx())
                                
  sgs = list(rwmol.GetStereoGroups())

  new_sgs = []
  for atom in rwmol.GetAtoms():
    if atom.HasProp("_ChiralityPossible"):
      tag = atom.GetChiralTag()
      if tag == Chem.ChiralType.CHI_UNSPECIFIED:
        if determine_unassigned_stereo:
          atom.SetChiralTag(Chem.ChiralType.CHI_TETRAHEDRAL_CW)
          new_sgs.append(Chem.CreateStereoGroup(Chem.StereoGroupType.STEREO_AND,rwmol,[atom.GetIdx()]))
      elif tag in [Chem.ChiralType.CHI_TETRAHEDRAL_CCW, Chem.ChiralType.CHI_TETRAHEDRAL_CW]:
        if atom.GetIdx() not in atoms_in_stereo_groups:
          new_sgs.append(Chem.CreateStereoGroup(Chem.StereoGroupType.STEREO_ABSOLUTE,rwmol,[atom.GetIdx()]))

    #elif atom.GetChiralTag() == None:
    #  asdf
  if new_sgs:
    rwmol.SetStereoGroups(sgs + new_sgs)
    return rwmol

  return mol

def applyStereoGroups(mol):
    """Apply existing stereogroups if possible, current applications
     STEREO_OR - ignored
     STEREO_ABSOLUTE - ignored
     STEREO_AND - if stereo group has a single atom, flatten the chirality
                 (make chirality unspecified)
    """
    rwmol = Chem.RWMol(mol)
    sgs = []
    for sg in rwmol.GetStereoGroups():
        if sg.GetGroupType() == Chem.StereoGroupType.STEREO_AND:
            atoms = list(sg.GetAtoms())
            if len(atoms) == 1:
                atoms[0].SetChiralTag(Chem.ChiralType.CHI_UNSPECIFIED)
            else:
                logging.warning(f"STEREO_AND StereoGroup {sg.GetReadId()} has more than one atom, ignoring")
                sgs.append(sg)
        elif sg.GetGroupType() == Chem.StereoGroupType.STEREO_OR:
                logging.warning(f"ignoring STEREO_OR for stereo group {sg.GetReadId()}")
                sgs.append(sg)            
        elif sg.GetGroupType == Chem.StereoGroupType.STEREO_ABSOLUTE:
            sgs.append(sg)
        else:
            sgs.append(sg)  
    
    rwmol.SetStereoGroups(sgs)
    return rwmol

smiles = [
  "C[C@H](F)Cl",
  "C[CH](F)Cl",
  "C[C@@H](F)Cl |&1:1|",
  "C[C@@H](F)Cl |o1:1|",
  "C[C@@H](F)Cl.C[C@@H](F)Cl |&1:1,&2:5|",
  "C[C@@H](F)Cl.C[C@@H](F)Cl |&1:1,o2:5|",
  ]

for smi in smiles:
  m = determineStereoGroups(Chem.MolFromSmiles(smi))
  m2 = applyStereoGroups(m)
  print(f"{smi} => determine sg {Chem.MolToCXSmiles(m)!r}  => apply sg {Chem.MolToSmiles(m2)!r}")
```

Results

```
C[C@H](F)Cl => determine sg 'C[C@H](F)Cl |a:1|'  => apply sg 'C[C@H](F)Cl'
C[CH](F)Cl => determine sg 'C[C@H](F)Cl |&1:1|'  => apply sg 'CC(F)Cl'
C[C@@H](F)Cl |&1:1| => determine sg 'C[C@H](F)Cl |&1:1|'  => apply sg 'CC(F)Cl'
WARNING:root:ignoring STEREO_OR for stereo group 1
C[C@@H](F)Cl |o1:1| => determine sg 'C[C@H](F)Cl |o1:1|'  => apply sg 'C[C@@H](F)Cl'
C[C@@H](F)Cl.C[C@@H](F)Cl |&1:1,&2:5| => determine sg 'C[C@H](F)Cl.C[C@H](F)Cl |&1:1,&2:5|'  => apply sg 'CC(F)Cl.CC(F)Cl'
WARNING:root:ignoring STEREO_OR for stereo group 2
C[C@@H](F)Cl.C[C@@H](F)Cl |&1:1,o2:5| => determine sg 'C[C@H](F)Cl.C[C@H](F)Cl |o1:5,&1:1|'  => apply sg 'CC(F)Cl.C[C@@H](F)Cl'
```

**Reply by bp-kelley** [reply]
Is this what you are talking about?

**Reply by bp-kelley** [reply]
You don't need to remove stereogroups, just don't output CXSmiles, so the only real operation is setting CHI_UNSPECIFIED for STEREO_AND groups with 1 atom

**Reply by bp-kelley** [reply]
I'm at a bit of a loss why this is useful though, there are so many cases it doesn't handle.

**Reply by pechersky** [reply]
One application for enhanced->traditional conversion is depiction. Depicting with `hash+&1` can be somewhat difficult to spot, especially if you have two racemic stereocenters, then you have to notice that the `&1` and `&2` are referring to different numbers. Versus the same number -- very different mixture. Additionally, "canonicalizing" the depiction to a flat bond might be better than not canonicalizing at all (hash+&1 and wedge+&1 look different, but are really depicting the same mixture), or canonicalizing to a preferred lets say hash+&1.

**Reply by bp-kelley** [reply]
CXSmiles already canonicalizes (hash+&1 and wedge+&1)

**Reply by pechersky** [reply]
Yes, depiction code paths need to know, "ah I must do ROMol -> CXSmiles -> ROMol -> depict, not just ROMol -> depict". And OE toolkits don't canonicalize CXSMILES =C

**Reply by bp-kelley** [reply]
Would it help if these were options in the depictor system?

**Reply by bp-kelley** [reply]
I.e. Render assumed enhance stereo?
